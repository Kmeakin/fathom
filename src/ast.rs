//! The syntax of our data description language

use source::Span;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Kind {
    Type,
}

/// A type definition
///
/// ```plain
/// Point = {
///     x : u16,
///     y : u16,
/// }
/// ```
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Definition {
    pub span: Span,
    pub name: String,
    pub ty: Type,
}

impl Definition {
    pub fn new<Sp, S>(span: Sp, name: S, ty: Type) -> Definition
    where
        Sp: Into<Span>,
        S: Into<String>,
    {
        let span = span.into();
        let name = name.into();

        Definition { span, name, ty }
    }
}

/// An expression
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expr {
    /// An integer constant: `0`, `1`, `2`, ...
    Const(Span, u64),
    /// A variable, referring to an integer that exists in the current context: `len`, `num_tables`
    Var(Span, String),
    /// Integer negation: `-x`
    Neg(Span, Box<Expr>),
    /// Integer addition: `x + y`
    Add(Span, Box<Expr>, Box<Expr>),
    /// Integer subtraction: `x - y`
    Sub(Span, Box<Expr>, Box<Expr>),
    /// Integer multiplication: `x * y`
    Mul(Span, Box<Expr>, Box<Expr>),
    /// Integer division: `x / y`
    Div(Span, Box<Expr>, Box<Expr>),
}

impl Expr {
    /// An integer constant: `0`, `1`, `2`, ...
    pub fn const_<Sp>(span: Sp, value: u64) -> Expr
    where
        Sp: Into<Span>,
    {
        Expr::Const(span.into(), value)
    }

    /// A variable, referring to an integer that exists in the current context: `len`, `num_tables`
    pub fn var<Sp, S>(span: Sp, name: S) -> Expr
    where
        Sp: Into<Span>,
        S: Into<String>,
    {
        Expr::Var(span.into(), name.into())
    }

    /// Integer negation: `-x`
    pub fn neg<Sp, T>(span: Sp, x: T) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
    {
        Expr::Neg(span.into(), x.into())
    }

    /// Integer addition: `x + y`
    pub fn add<Sp, T, U>(span: Sp, x: T, y: U) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
        U: Into<Box<Expr>>,
    {
        Expr::Add(span.into(), x.into(), y.into())
    }

    /// Integer subtraction: `x - y`
    pub fn sub<Sp, T, U>(span: Sp, x: T, y: U) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
        U: Into<Box<Expr>>,
    {
        Expr::Sub(span.into(), x.into(), y.into())
    }

    /// Integer multiplication: `x * y`
    pub fn mul<Sp, T, U>(span: Sp, x: T, y: U) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
        U: Into<Box<Expr>>,
    {
        Expr::Mul(span.into(), x.into(), y.into())
    }

    /// Integer division: `x / y`
    pub fn div<Sp, T, U>(span: Sp, x: T, y: U) -> Expr
    where
        Sp: Into<Span>,
        T: Into<Box<Expr>>,
        U: Into<Box<Expr>>,
    {
        Expr::Div(span.into(), x.into(), y.into())
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum TypeConst {
    U8,
    U16,
    U32,
    U64,
    I8,
    I16,
    I32,
    I64,
    F32,
    F64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    /// A type constant
    Const(Span, TypeConst),
    /// A type identifier: `T1`, `u16be`, `u32`
    Ident(Span, String),
    /// An array of the specified type, with a size: `[T; n]`
    Array(Span, Box<Type>, Expr),
    /// A union of types: `union { T, ... }`
    Union(Span, Vec<Type>),
    /// A struct type, with fields: `struct { field : T, ... }`
    Struct(Span, Vec<Field>),
}

impl Type {
    /// A type constant
    pub fn const_<Sp>(span: Sp, ty: TypeConst) -> Type
    where
        Sp: Into<Span>,
    {
        Type::Const(span.into(), ty)
    }

    /// A type identifier: `T1`, `u16be`, `u32`
    pub fn ident<Sp, S>(span: Sp, name: S) -> Type
    where
        Sp: Into<Span>,
        S: Into<String>,
    {
        Type::Ident(span.into(), name.into())
    }

    /// An array of the specified type, with a size: `[T; n]`
    pub fn array<Sp, T>(span: Sp, ty: T, size: Expr) -> Type
    where
        Sp: Into<Span>,
        T: Into<Box<Type>>,
    {
        Type::Array(span.into(), ty.into(), size)
    }

    /// A union of types: `union { T, ... }`
    pub fn union<Sp>(span: Sp, tys: Vec<Type>) -> Type
    where
        Sp: Into<Span>,
    {
        Type::Union(span.into(), tys)
    }

    /// A struct type, with fields: `struct { field : T, ... }`
    pub fn struct_<Sp>(span: Sp, fields: Vec<Field>) -> Type
    where
        Sp: Into<Span>,
    {
        Type::Struct(span.into(), fields)
    }
}

/// A field in a struct type
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Field {
    pub span: Span,
    pub name: String,
    pub ty: Type,
}

impl Field {
    pub fn new<Sp, S>(span: Sp, name: S, ty: Type) -> Field
    where
        Sp: Into<Span>,
        S: Into<String>,
    {
        let span = span.into();
        let name = name.into();

        Field { span, name, ty }
    }
}
