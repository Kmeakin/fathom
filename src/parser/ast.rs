//! The parse tree of the DDL
//!
//! This has a 1:1 mapping to the exposed concrete syntax. In order to do
//! anything interesting with this, one must first lower it to the core
//! representation.

use codespan::{BytePos, Span};

use parser::{self, grammar, ParseError};
use parser::lexer::Lexer;

pub use syntax::ast::{Binop, Const, FloatType, IntSuffix, TypeConst, Unop};

#[derive(Debug, Clone, PartialEq)]
pub struct Module<'input> {
    pub definitions: Vec<Definition<'input>>,
}

impl<'input> Module<'input> {
    /// Attempt to parse a module from a source string
    pub fn from_str(src: &'input str) -> Result<Module<'input>, ParseError> {
        grammar::parse_Module(Lexer::new(src).map(|x| x.map_err(ParseError::from)))
            .map_err(|err| parser::from_lalrpop(src, err))
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Definition<'input> {
    pub doc: Vec<&'input str>,
    pub span: Span,
    pub name: &'input str,
    pub param_names: Vec<&'input str>,
    pub body_ty: Type<'input>,
}

impl<'input> Definition<'input> {
    /// Attempt to parse a definition from a source string
    pub fn from_str(src: &'input str) -> Result<Definition<'input>, ParseError> {
        grammar::parse_Definition(Lexer::new(src).map(|x| x.map_err(ParseError::from)))
            .map_err(|err| parser::from_lalrpop(src, err))
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Field<'input, T> {
    pub doc: Vec<&'input str>,
    pub name: &'input str,
    pub value: T,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Type<'input> {
    Var(Span, &'input str),
    Array(Span, Box<Type<'input>>, Box<Expr<'input>>),
    Cond(Span, Vec<Field<'input, (Expr<'input>, Type<'input>)>>),
    Struct(Span, Vec<Field<'input, Type<'input>>>),
    Where(
        Span,
        Box<Type<'input>>,
        BytePos,
        &'input str,
        Box<Expr<'input>>,
    ),
    Compute(Span, TypeConst, Box<Expr<'input>>),
    App(Span, Box<Type<'input>>, Vec<Type<'input>>),
}

impl<'input> Type<'input> {
    /// Attempt to parse a type from a source string
    pub fn from_str(src: &'input str) -> Result<Type<'input>, ParseError> {
        grammar::parse_Type(Lexer::new(src).map(|x| x.map_err(ParseError::from)))
            .map_err(|err| parser::from_lalrpop(src, err))
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr<'input> {
    Ann(Span, Box<Expr<'input>>, TypeConst),
    Const(Span, Const),
    Var(Span, &'input str),
    Unop(Span, Unop, Box<Expr<'input>>),
    Binop(Span, Binop, Box<Expr<'input>>, Box<Expr<'input>>),
    Array(Span, Vec<Expr<'input>>),
    Proj(Span, Box<Expr<'input>>, &'input str),
    Subscript(Span, Box<Expr<'input>>, Box<Expr<'input>>),
    Cast(Span, Box<Expr<'input>>, TypeConst),
}

impl<'input> Expr<'input> {
    /// Attempt to parse an expression from a source string
    pub fn from_str(src: &'input str) -> Result<Expr<'input>, ParseError> {
        grammar::parse_Expr(Lexer::new(src).map(|x| x.map_err(ParseError::from)))
            .map_err(|err| parser::from_lalrpop(src, err))
    }
}
