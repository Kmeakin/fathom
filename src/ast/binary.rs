//! The syntax of our data description language

use source::Span;
use ast::{host, Field, Named, Var};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Kind {
    Type,
    Arrow(Box<Kind>, Box<Kind>),
}

impl Kind {
    pub fn arrow<T, U>(lhs: T, rhs: U) -> Kind
    where
        T: Into<Box<Kind>>,
        U: Into<Box<Kind>>,
    {
        Kind::Arrow(lhs.into(), rhs.into())
    }
}

/// A binary type
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TypeF<N, T, E> {
    /// A type variable: eg. `T`
    Var(Var<N, Named<N, u32>>),
    /// Bit
    Bit,
    /// An array of the specified type, with a size: eg. `[T; n]`
    Array(Box<T>, Box<E>),
    /// A union of types: eg. `union { T, ... }`
    Union(Vec<T>),
    /// A struct type, with fields: eg. `struct { field : T, ... }`
    Struct(Vec<Field<N, T>>),
    /// A type constrained by a predicate: eg. `T where x => x == 3`
    Cond(Named<N, Box<T>>, Box<E>),
    /// Type abstraction: eg. `\(a : Type) -> T`
    Abs(Vec<Named<N, Kind>>, Box<T>),
    /// Type application: eg. `T U V`
    App(Box<T>, Vec<T>),
}

/// A binary type AST
pub trait TypeNode<N, E: host::ExprNode<N>>: Sized + AsRef<TypeF<N, Self, E>> + AsMut<TypeF<N, Self, E>> {}

/// The recursive innards of a `Type`
///
/// This does the job of tying the recursive knot for `Type`, turning
/// `TypeF` into a tree of expressions.
///
/// We could have just inlined this inside `Type`, but this type alias is
/// handy to have around for defining our conversion impls.
pub type TypeRec<N> = TypeF<N, Type<N>, host::Expr<N>>;

/// A tree of binary types
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Type<N>(pub TypeRec<N>);

impl<N> TypeNode<N, host::Expr<N>> for Type<N> {}

impl<N> Into<TypeRec<N>> for Type<N> {
    fn into(self) -> TypeRec<N> {
        self.0
    }
}

impl<N> AsRef<TypeRec<N>> for Type<N> {
    fn as_ref(&self) -> &TypeRec<N> {
        &self.0
    }
}

impl<N> AsMut<TypeRec<N>> for Type<N> {
    fn as_mut(&mut self) -> &mut TypeRec<N> {
        &mut self.0
    }
}

/// The recursive innards of a `SpannedType`
///
/// This does the job of tying the recursive knot for `SpannedType`, turning
/// `TypeF` into a tree of spanned expressions.
///
/// We could have just inlined this inside `SpannedType`, but this type alias is
/// handy to have around for defining our conversion impls.
pub type SpannedTypeRec<N> = TypeF<N, SpannedType<N>, host::SpannedExpr<N>>;

/// A tree of spanned binary types
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SpannedType<N> {
    pub span: Span,
    pub inner: SpannedTypeRec<N>,
}

impl<N> TypeNode<N, host::SpannedExpr<N>> for SpannedType<N> {}

impl<N> Into<SpannedTypeRec<N>> for SpannedType<N> {
    fn into(self) -> SpannedTypeRec<N> {
        self.inner
    }
}

impl<N> AsRef<SpannedTypeRec<N>> for SpannedType<N> {
    fn as_ref(&self) -> &SpannedTypeRec<N> {
        &self.inner
    }
}

impl<N> AsMut<SpannedTypeRec<N>> for SpannedType<N> {
    fn as_mut(&mut self) -> &mut SpannedTypeRec<N> {
        &mut self.inner
    }
}

impl<N, T, E> TypeF<N, T, E> {
    /// A free type variable: eg. `T`
    pub fn fvar<N1: Into<N>>(x: N1) -> TypeF<N, T, E> {
        TypeF::Var(Var::Free(x.into()))
    }

    /// A bound type variable
    pub fn bvar<N1: Into<N>>(x: N1, i: u32) -> TypeF<N, T, E> {
        TypeF::Var(Var::Bound(Named(x.into(), i)))
    }

    /// An array of the specified type, with a size: eg. `[T; n]`
    pub fn array<T1: Into<Box<T>>, E1: Into<Box<E>>>(ty: T1, size: E1) -> TypeF<N, T, E> {
        TypeF::Array(ty.into(), size.into())
    }

    /// A union of types: eg. `union { T, ... }`
    pub fn union(tys: Vec<T>) -> TypeF<N, T, E> {
        TypeF::Union(tys)
    }
}

impl<N, T, E> TypeF<N, T, E>
where
    T: TypeNode<N, E>,
    E: host::ExprNode<N>,
{
    /// A struct type, with fields: eg. `struct { field : T, ... }`
    pub fn struct_<Fs>(fields: Fs) -> TypeF<N, T, E>
    where
        N: PartialEq + Clone,
        Fs: IntoIterator<Item = Field<N, T>>,
    {
        // We maintain a list of the seen field names. This will allow us to
        // recover the index of these variables as we abstract later fields...
        let mut seen_names = Vec::new();
        let f = |field: Field<N, T>| {
            let field = field.map_value(|mut ty| {
                T::as_mut(&mut ty).abstract_with(&|x| {
                    seen_names
                        .iter()
                        .position(|y| x == y)
                        .map(|i| Named(x.clone(), i as u32))
                });
                ty
            });

            // Record that the field has been 'seen'
            seen_names.push(field.name.clone());

            field
        };

        TypeF::Struct(fields.into_iter().map(f).collect())
    }

    /// A type constrained by a predicate: eg. `T where x => x == 3`
    pub fn cond<N1, T1, E1>(ty: T1, param: N1, pred: E1) -> TypeF<N, T, E>
    where
        N: PartialEq + Clone,
        N1: Into<N>,
        T1: Into<Box<T>>,
        E1: Into<Box<E>>,
    {
        let param = param.into();
        let mut pred = pred.into();
        E::as_mut(&mut pred).abstract_name(&param);

        TypeF::Cond(Named(param, ty.into()), pred)
    }

    /// Type abstraction: eg. `\(a : Type) -> T`
    pub fn abs<T1>(params: Vec<Named<N, Kind>>, ty: T1) -> TypeF<N, T, E>
    where
        N: PartialEq + Clone,
        T1: Into<Box<T>>,
    {
        let mut ty = ty.into();
        T::as_mut(&mut ty).abstract_with(&|x| {
            params
                .iter()
                .position(|&Named(ref y, _)| x == y)
                .map(|i| Named(x.clone(), i as u32))
        });

        TypeF::Abs(params, ty)
    }

    /// Type application: eg. `T U V`
    pub fn app<T1: Into<Box<T>>>(ty: T1, args: Vec<T>) -> TypeF<N, T, E> {
        TypeF::App(ty.into(), args)
    }

    /// `true` if the type contains no free variables in itself or its sub-expressions
    pub fn is_closed(&self) -> bool {
        match *self {
            TypeF::Var(ref v) => v.is_closed(),
            TypeF::Bit => true,
            TypeF::Array(ref t, ref e) => T::as_ref(t).is_closed() && E::as_ref(e).is_closed(),
            TypeF::Union(ref ts) => ts.iter().all(|t| T::as_ref(t).is_closed()),
            TypeF::Struct(ref fs) => fs.iter().all(|f| T::as_ref(&f.value).is_closed()),
            TypeF::Cond(Named(_, ref t), ref e) => {
                T::as_ref(t).is_closed() && E::as_ref(e).is_closed()
            }
            TypeF::Abs(_, ref t) => T::as_ref(t).is_closed(),
            TypeF::App(ref t, ref ts) => {
                T::as_ref(t).is_closed() && ts.iter().all(|t| T::as_ref(t).is_closed())
            }
        }
    }

    fn abstract_level_with<F>(&mut self, level: u32, f: &F)
    where
        F: Fn(&N) -> Option<Named<N, u32>>,
    {
        match *self {
            TypeF::Var(ref mut v) => v.abstract_with(f),
            TypeF::Bit => {}
            TypeF::Array(ref mut ty, ref mut e) => {
                T::as_mut(ty).abstract_level_with(level, f);
                E::as_mut(e).abstract_level_with(level, f);
            }
            TypeF::Cond(Named(_, ref mut ty), ref mut e) => {
                T::as_mut(ty).abstract_level_with(level, f);
                E::as_mut(e).abstract_level_with(level, f);
            }
            TypeF::Abs(ref params, ref mut ty) => {
                T::as_mut(ty).abstract_level_with(level + params.len() as u32, f);
            }
            TypeF::Union(ref mut tys) => for ty in tys {
                T::as_mut(ty).abstract_level_with(level, f);
            },
            TypeF::Struct(ref mut fs) => for (i, field) in fs.iter_mut().enumerate() {
                T::as_mut(&mut field.value).abstract_level_with(level + i as u32, f);
            },
            TypeF::App(ref mut ty, ref mut tys) => {
                T::as_mut(ty).abstract_level_with(level, f);
                for ty in tys {
                    T::as_mut(ty).abstract_level_with(level, f);
                }
            }
        };
    }

    pub fn abstract_with<F>(&mut self, f: &F)
    where
        F: Fn(&N) -> Option<Named<N, u32>>,
    {
        self.abstract_level_with(0, &f);
    }

    pub fn abstract_name(&mut self, x: &N)
    where
        N: PartialEq + Clone,
    {
        self.abstract_with(&|y| if x == y {
            Some(Named(x.clone(), 0))
        } else {
            None
        });
    }

    fn instantiate_level(&mut self, level: u32, src: &TypeF<N, T, E>)
    where
        N: Clone,
        T: Clone,
        E: Clone,
    {
        // Bleh: Running into non-lexical liftetime problems here!
        // Just so you know that I'm not going completely insane....
        // FIXME: ensure that expressions are not bound at the same level
        *self = match *self {
            TypeF::Var(Var::Bound(ref i)) => if *i == level {
                src.clone()
            } else {
                return;
            },
            TypeF::Var(Var::Free(_)) | TypeF::Bit => return,
            TypeF::Array(ref mut ty, _) => {
                T::as_mut(ty).instantiate_level(level, src);
                return;
            }
            TypeF::Cond(Named(_, ref mut ty), _) => {
                T::as_mut(ty).instantiate_level(level, src);
                return;
            }
            TypeF::Union(ref mut tys) => {
                for ty in tys {
                    T::as_mut(ty).instantiate_level(level, src);
                }
                return;
            }
            TypeF::Struct(ref mut fs) => {
                for (i, f) in fs.iter_mut().enumerate() {
                    T::as_mut(&mut f.value).instantiate_level(level + i as u32, src);
                }
                return;
            }
            TypeF::Abs(ref params, ref mut ty) => {
                T::as_mut(ty).instantiate_level(level + params.len() as u32, src);
                return;
            }
            TypeF::App(ref mut ty, ref mut tys) => {
                T::as_mut(ty).instantiate_level(level, src);
                for ty in tys {
                    T::as_mut(ty).instantiate_level(level, src);
                }
                return;
            }
        };
    }

    pub fn instantiate(&mut self, ty: &TypeF<N, T, E>)
    where
        N: Clone,
        T: Clone,
        E: Clone,
    {
        self.instantiate_level(0, ty);
    }
}
