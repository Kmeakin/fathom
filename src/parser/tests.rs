use syntax::ast::Module;
use syntax::ast::binary::RcType;
use syntax::ast::host::RcIExpr;
use parser::ast::Module as ParseModule;
use parser::ast::binary::Type as ParseType;
use parser::ast::host::Expr as ParseExpr;

#[test]
fn parse_expr_bool_atomic() {
    let src = "
        !((true || (false)))
    ";

    assert_debug_snapshot!(
        parse_expr_bool_atomic,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_expr_bool_operators() {
    let src = "
        (true && false) || (true || false)
    ";

    assert_debug_snapshot!(
        parse_expr_bool_operators,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_add_expr() {
    let src = "x + y + z";

    assert_debug_snapshot!(
        parse_add_expr,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_sub_expr() {
    let src = "x - y - z";

    assert_debug_snapshot!(
        parse_sub_expr,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_add_expr_mixed() {
    let src = "x + y + z - z + x";

    assert_debug_snapshot!(
        parse_add_expr_mixed,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_mul_expr() {
    let src = "x * y * z";

    assert_debug_snapshot!(
        parse_mul_expr,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_div_expr() {
    let src = "x / y / z";

    assert_debug_snapshot!(
        parse_div_expr,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_mul_expr_mixed() {
    let src = "x * y * z / z * x";

    assert_debug_snapshot!(
        parse_mul_expr_mixed,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_mixed_arithmetic_expr() {
    let src = "x + y * z / z - x * a";

    assert_debug_snapshot!(
        parse_mixed_arithmetic_expr,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_mixed_arithmetic_expr_parenthesized() {
    let src = "(x + y) * z / (z - x) * a";

    assert_debug_snapshot!(
        parse_mixed_arithmetic_expr_parenthesized,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_proj_expr() {
    let src = "-foo.bar.x";

    assert_debug_snapshot!(
        parse_proj_expr,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_subscript_expr() {
    let src = "-foo[23u32 + (2u32 + 3u32)][index]";

    assert_debug_snapshot!(
        parse_subscript_expr,
        RcIExpr::from_parse(&ParseExpr::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_ty_var() {
    let src = "
        Point
    ";

    assert_debug_snapshot!(
        parse_ty_var,
        RcType::from_parse(&ParseType::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_ty_empty_struct() {
    let src = "struct {}";

    assert_debug_snapshot!(
        parse_ty_empty_struct,
        RcType::from_parse(&ParseType::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_ty_assert() {
    let src = "
        struct {
            x: u32 where x => x == 3u32,
        }
        where x => x == 2u32
        where x => x == 1u32
    ";

    assert_debug_snapshot!(
        parse_ty_assert,
        RcType::from_parse(&ParseType::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_ty_array_dependent() {
    let src = "
        struct {
            len: u32,
            data1: [f32; len],
            data2: struct {
                data1: [f32; len],
                len: u32,
                /// Doc comment
                padding1: u8,
                padding2: u8,
                data2: [f32; len],
            },
        }
    ";

    assert_debug_snapshot!(
        parse_ty_array_dependent,
        RcType::from_parse(&ParseType::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_simple_definition() {
    let src = "
        Offset32 = u32;
    ";

    assert_debug_snapshot!(
        parse_simple_definition,
        Module::from_parse(&ParseModule::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_array_with_constant_size() {
    let src = "
        Point = [f32; 3u32];
    ";

    assert_debug_snapshot!(
        parse_array_with_constant_size,
        Module::from_parse(&ParseModule::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_definition() {
    let src = "
        Point = struct {
            x : u32be,
            y : u32be,
        };

        /// A doc comment
        ///
        /// With some mor text
        Array = struct {
            len : u16le,
            data : [Point; len],
        };

        Formats = struct {
            format : u8,
            data : cond {
                /// A doc comment on a union variant
                u16 : format == 0x00u8 => u16,
                point : format == 0x01u8 => Point,
                /// And another doc comment
                array : format == 0x02u8 => Array,
            },
        };
    ";

    assert_debug_snapshot!(
        parse_definition,
        Module::from_parse(&ParseModule::from_str(src).unwrap()).unwrap()
    );
}

#[test]
fn parse_type_params() {
    let src = "
        Pair(T, U) = struct {
            l : T,
            r : U,
        };

        Array(T) = struct {
            len : u16le,
            data : [T; len],
        };

        Data(T, U, V) = struct {
            blah : U,
            data : Array(Pair(T, V)),
        };
    ";

    assert_debug_snapshot!(
        parse_type_params,
        Module::from_parse(&ParseModule::from_str(src).unwrap()).unwrap()
    );
}
