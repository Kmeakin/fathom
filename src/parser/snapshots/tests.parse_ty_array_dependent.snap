Struct(
    Span {
        lo: BytePos(9),
        hi: BytePos(311)
    },
    [
        Field {
            doc: "",
            name: OwnedIdent("len"),
            value: Var(
                Span {
                    lo: BytePos(35),
                    hi: BytePos(38)
                },
                Free(OwnedIdent("u32"))
            )
        },
        Field {
            doc: "",
            name: OwnedIdent("data1"),
            value: Array(
                Span {
                    lo: BytePos(59),
                    hi: BytePos(69)
                },
                Var(
                    Span {
                        lo: BytePos(60),
                        hi: BytePos(63)
                    },
                    Free(OwnedIdent("f32"))
                ),
                Var(
                    Span {
                        lo: BytePos(65),
                        hi: BytePos(68)
                    },
                    Bound(Named(OwnedIdent("len"), BoundVar(0, 0)))
                )
            )
        },
        Field {
            doc: "",
            name: OwnedIdent("data2"),
            value: Struct(
                Span {
                    lo: BytePos(90),
                    hi: BytePos(300)
                },
                [
                    Field {
                        doc: "",
                        name: OwnedIdent("data1"),
                        value: Array(
                            Span {
                                lo: BytePos(122),
                                hi: BytePos(132)
                            },
                            Var(
                                Span {
                                    lo: BytePos(123),
                                    hi: BytePos(126)
                                },
                                Free(OwnedIdent("f32"))
                            ),
                            Var(
                                Span {
                                    lo: BytePos(128),
                                    hi: BytePos(131)
                                },
                                Bound(Named(OwnedIdent("len"), BoundVar(1, 0)))
                            )
                        )
                    },
                    Field {
                        doc: "",
                        name: OwnedIdent("len"),
                        value: Var(
                            Span {
                                lo: BytePos(155),
                                hi: BytePos(158)
                            },
                            Free(OwnedIdent("u32"))
                        )
                    },
                    Field {
                        doc: "Doc comment",
                        name: OwnedIdent("padding1"),
                        value: Var(
                            Span {
                                lo: BytePos(218),
                                hi: BytePos(220)
                            },
                            Free(OwnedIdent("u8"))
                        )
                    },
                    Field {
                        doc: "",
                        name: OwnedIdent("padding2"),
                        value: Var(
                            Span {
                                lo: BytePos(248),
                                hi: BytePos(250)
                            },
                            Free(OwnedIdent("u8"))
                        )
                    },
                    Field {
                        doc: "",
                        name: OwnedIdent("data2"),
                        value: Array(
                            Span {
                                lo: BytePos(275),
                                hi: BytePos(285)
                            },
                            Var(
                                Span {
                                    lo: BytePos(276),
                                    hi: BytePos(279)
                                },
                                Free(OwnedIdent("f32"))
                            ),
                            Var(
                                Span {
                                    lo: BytePos(281),
                                    hi: BytePos(284)
                                },
                                Bound(Named(OwnedIdent("len"), BoundVar(2, 0)))
                            )
                        )
                    }
                ]
            )
        }
    ]
)
